src/node_file.cc

```diff
@@ -3431,6 +3432,27 @@ static void CpSyncCopyDir(const FunctionCallbackInfo<Value>& args) {
   bool verbatim_symlinks = args[5]->IsTrue();
   bool preserve_timestamps = args[6]->IsTrue();
 
+  std::optional<std::function<bool(std::string, std::string)>> filter_fn;
+
+  if (args[7]->IsFunction()) {
+    Local<v8::Function> args_filter_fn = args[7].As<v8::Function>();
+
+    filter_fn = [env, args_filter_fn](std::string_view src,
+                                      std::string_view dest) -> bool {
+      Local<Value> argv[] = {
+          String::NewFromUtf8(
+              env->isolate(), src.data(), v8::NewStringType::kNormal)
+              .ToLocalChecked(),
+          String::NewFromUtf8(
+              env->isolate(), dest.data(), v8::NewStringType::kNormal)
+              .ToLocalChecked()};
```

Comment:
One additional difference here that will be problematic is that this is passing the `src` and `dest` to the filter callback as a string always. With the currrent implementation, however, if we pass the pass in as a Buffer or URL then the filter callback receives the Buffer and URL... I notice that this appears to be undocumented in api docs sadly)... We have to preserve that because this implementation tries to interpret all src and dest paths as UTF8 sequences which they quite likely won't be (the reason we accept Buffer in the first place is that many filenames in older legacy systems using legacy text encodings).

```
fs.cpSync(Buffer.from('a.js'), Buffer.from('b.js'), { filter(...args) { console.log(...args); return true } });
```

Reply:
Ah ok I see

Thanks a lot for pointing this out, I'll address it :pray:

Reply:
@jasnell Are you sure about this statement?

> if we pass the path in as a Buffer or URL then the filter callback receives the Buffer and URL...

Because after giving it a quick try that does not seem to be the case for me :thinking: 
![Screenshot at 2025-06-08 12-38-11](https://github.com/user-attachments/assets/2b1233a0-4545-4d6f-b14d-108958e1b445)

Reply:
I've also commented this here: https://github.com/nodejs/node/pull/58627#pullrequestreview-2908421074

Reply:
The reason is simple enough... I'd like to make sure that the fix to the bug doesn't conflict with the changes you're making here... that is, I want to make sure landing this doesn't make fixing the bug more difficult / convoluted / etc or make sure we don't end up having to revert this in part when fixing the issue. Even if we just can get a sense of how we want to fix the bug it should be enough to unblock this.